rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -532,3 +569,164 @@ class YieldExprTree extends StandardPostOrderTree instanceof YieldExpr {
+        first(this.getPat(i + 1), succ)
+        or
+        // Edge from successful pattern to this
+        c.(MatchCompletion).succeeded() and
+        succ = this
+        or
+        // Edge from last pattern to this
+        not exists(this.getPat(i + 1)) and
+        succ = this and
+        completionIsNormal(c)
+      )
+    }
+  }
+
+  class IdentPatTree extends PostOrderPatTree, IdentPat {
+    override Pat getPat(int i) { i = 0 and result = this.getPat() }
+
+    override predicate last(AstNode node, Completion c) {
+      super.last(node, c)
+      or
+      last(this.getPat(), node, c) and
+      c.(MatchCompletion).failed()
+    }
+
+    override predicate succ(AstNode pred, AstNode succ, Completion c) {
+      super.succ(pred, succ, c) and
+      not (succ = this and c.(MatchCompletion).failed())
+    }
+  }
+
+  class BoxPatTree extends PreOrderPatTree, BoxPat {
+    override Pat getPat(int i) { i = 0 and result = this.getPat() }
+  }
+
+  class RestPatTree extends LeafTree, RestPat { }
+
+  class LiteralPatTree extends LeafTree, LiteralPat { }
+
+  class MacroPatTree extends LeafTree, MacroPat { } // todo
+
+  class OrPatTree extends PostOrderPatTree instanceof OrPat {
+    override Pat getPat(int i) { result = OrPat.super.getPat(i) }
+  }
+
+  class ParenPatTree extends ControlFlowTree, ParenPat {
+    private ControlFlowTree pat;
+
+    ParenPatTree() { pat = this.getPat() }
+
+    override predicate propagatesAbnormal(AstNode child) { pat.propagatesAbnormal(child) }
+
+    override predicate first(AstNode first) { pat.first(first) }
+
+    override predicate last(AstNode last, Completion c) { pat.last(last, c) }
+
+    override predicate succ(AstNode pred, AstNode succ, Completion c) { none() }
+  }
+
+  class PathPatTree extends LeafTree, PathPat { }
+
+  class WildcardPatTree extends LeafTree, WildcardPat { }
+
+  class RangePatTree extends PreOrderPatTree, RangePat {
+    override Pat getPat(int i) {
+      i = 0 and result = this.getStart()
+      or
+      i = 1 and result = this.getEnd()
+    }
+  }
+
+  class RecordPatTree extends PreOrderPatTree, RecordPat {
+    override Pat getPat(int i) {
+      result = this.getRecordPatFieldList().getField(i).getPat()
+      or
+      i = this.getRecordPatFieldList().getNumberOfFields() and
+      result = this.getRecordPatFieldList().getRestPat()
+    }
+  }
+
+  class RefPatTree extends PreOrderPatTree, RefPat {
+    override Pat getPat(int i) { i = 0 and result = super.getPat() }
+  }
+
+  class SlicePatTree extends PreOrderPatTree instanceof SlicePat {
+    override Pat getPat(int i) { result = SlicePat.super.getPat(i) }
+  }
+
+  class TuplePatTree extends PreOrderPatTree, TuplePat {
+    override Pat getPat(int i) { result = this.getField(i) }
+  }
+
+  class TupleStructPatTree extends PreOrderPatTree, TupleStructPat {
+    override Pat getPat(int i) { result = this.getField(i) }
+  }
+
+  class ConstBlockPatTree extends LeafTree, ConstBlockPat { } // todo?
```

Comment:
I think the constant value inside is evaluated compile-time, so perhaps having this as a `LeafTree` is the right thing to do.